src/api/sync.rs

```diff
@@ -270,12 +264,51 @@ impl Api {
     }
 
     fn metadata(&self, url: &str) -> Result<Metadata, ApiError> {
-        let response = self
+        let mut response = self
             .no_redirect_client
             .get(url)
             .set(RANGE, "bytes=0-0")
             .call()
             .map_err(Box::new)?;
+
+        // See: https://github.com/huggingface/huggingface_hub/blob/9c6af39cdce45b570f0b7f8fad2b311c96019804/src/huggingface_hub/file_download.py#L411
+
+        // Closure to check if status code is a redirection
+        let should_redirect = |status_code: u16| {
+            matches!(
+                StatusCode::from_u16(status_code).unwrap(),
+                StatusCode::MOVED_PERMANENTLY
+                    | StatusCode::FOUND
+                    | StatusCode::SEE_OTHER
+                    | StatusCode::TEMPORARY_REDIRECT
+                    | StatusCode::PERMANENT_REDIRECT
+            )
+        };
+
+        // Follow redirects until host does not match if the first host
+        let first_uri = Uri::from_str(response.get_url()).unwrap();
+        let response = loop {
+            // Check status
+            if should_redirect(response.status()) {
+                // Get header value
+                if let Some(location) = response.header("Location") {
+                    let url = Uri::from_str(location).unwrap();
+                    // Match hosts
+                    if url.host() == first_uri.host() {
+                        // Follow redirect
+                        response = self
+                            .no_redirect_client
+                            .get(response.get_url())
+                            .set(RANGE, "bytes=0-0")
```

Comment:
do you happen to know why the GET range 0-0 instead of a HEAD?

Reply:
Because specific DNS routing rules with headers.
Although I'm not 100% sure if it's still needed.